crates/matrix-sdk/src/widget_api/driver/mod.rs

```diff
@@ -106,126 +114,78 @@ impl handler::EventReader for EventReader {
+            events: state
+                .chain(timeline)
+                .filter_map(StdResult::ok)
+                .filter(|m| self.filter.allow(m))
+                .collect(),
+        })
     }
 }
-#[derive(Debug)]
-pub struct EventSender {
-    room: Joined,
-    filter: Vec<Filter>,
-}
+
 #[async_trait]
-impl handler::EventSender for EventSender {
-    fn get_filter(&self) -> &Vec<Filter> {
-        &self.filter
-    }
+impl Sender for EventHandler {
     async fn send(&self, req: SendEventRequest) -> Result<SendEventResponse> {
-        let filter_fn = |f: &Filter| f.allow_event(&req.message_type, &req.state_key, &req.content);
-
-        if self.filter.iter().any(filter_fn) {
-            match req.state_key {
-                Some(state_key) => {
-                    match self
-                        .room
-                        .send_state_event_raw(req.content, &req.message_type, &state_key)
-                        .await
-                    {
-                        Ok(send_res) => Ok(SendEventResponse {
-                            room_id: self.room.room_id().to_string(),
-                            event_id: send_res.event_id.to_string(),
-                        }),
-                        Err(err) => {
-                            Err(WidgetError(format!("Could not send event with error: {}", err)))
-                        }
-                    }
-                }
-                None => match self.room.send_raw(req.content, &req.message_type, None).await {
-                    Ok(send_res) => Ok(SendEventResponse {
-                        room_id: self.room.room_id().to_string(),
-                        event_id: send_res.event_id.to_string(),
-                    }),
-                    Err(err) => {
-                        Err(WidgetError(format!("Could not send event with error: {}", err)))
-                    }
-                },
-            }
-        } else {
-            Err(WidgetError(format!(
-                "No capability to send event of type {} with state key {} (for room events the state key is undefined if no state key is shown the state key is \"\")",
-                req.message_type, req.state_key.unwrap_or("undefined".to_string())
-            )))
+        // Run the request through the filter.
+        if !self.filter.allow(&req) {
+            return Err(Error::InvalidPermissions);
         }
+
+        // Send the requset based on whether the state key is set or not.
+        let event_id = match req.state_key {
+            Some(key) => {
+                self.room
+                    .send_state_event_raw(req.content, &req.message_type, &key)
+                    .await
+                    .map_err(|_| Error::Other)?
+                    .event_id
+            }
+            None => {
+                self.room
+                    .send_raw(req.content, &req.message_type, None)
+                    .await
+                    .map_err(|_| Error::Other)?
```

Comment:
I think it would be useful if we could make a return type `Result<T, String>` or something like this, this is what I meant with one of my TODOs in the description - error handling must be revisited. We would then forward the error to the widget (we don't do anything else with an error right now apart from just sending it back to the widget).